_posts/2023-04-19-more-optimizations.md

```diff
@@ -0,0 +1,1300 @@
+    {test,is_integer,{f,19},[{x,1}]}.
+    {test,is_ge,{f,19},[{tr,{x,1},{t_integer,any}},{integer,0}]}.
+    {test,is_ge,{f,19},[{integer,1023},{tr,{x,1},{t_integer,{0,'+inf'}}}]}.
+    {gc_bif,'+',
+            {f,0},
+            2,
+            [{tr,{x,0},{t_integer,{0,1023}}},{tr,{x,1},{t_integer,{0,1023}}}],
+            {x,0}}.
+```
+
+The BEAM instruction for the `+` operator now have ranges for its operands.
+
+Let's look at little bit closer at the first three instructions, which
+corresponds to the guard test `is_integer(X), 0 =< X, X < 16#400`.
+
+First is the guard check for an integer:
+
+```
+    {test,is_integer,{f,19},[{x,0}]}.
+```
+
+It is followed by the guard test `X =< 0`:
+
+```
+    {test,is_ge,{f,19},[{tr,{x,0},{t_integer,any}},{integer,0}]}.
+```
+
+As a result of the `is_integer/1` test, it is known that `{x,0}`
+is an integer.
+
+The third instruction corresponds to `X < 16#400`, which the compiler rewrote
+to `16#3FF =< X` (`1023 =< X`):
+
+```
+    {test,is_ge,{f,19},[{integer,1023},{tr,{x,0},{t_integer,{0,'+inf'}}}]}.
+```
+
+In the type for the `{x,0}` register there is something new for OTP
+26. It says that the range is 0 through `'+inf'`, that is, from 0 up
+to positive infinity. From that information, the compiler can infer
+that if this instruction succeeds, the type for `{x,0}` is
+`t_integer,{0,1023}}`.
+
+
+### Combining guard tests
+
+In OTP 25, the JIT would emit native code for each BEAM instruction
+in the guard individually. When translated individually, the three guards
+tests for one of the variables each require 11 native instructions, or 33
+instructions for all three.
+
+By having the BEAM loader combining the three guard tests into a
+single `is_int_range` instruction, the JIT is capable of doing a much
+better job and emit a mere 6 native instructions.
+
+How is that possible?
+
+As individual BEAM instructions, each guard test needs 5 instructions
+to fetch the value from `{x,0}` and test that the value is a small
+integer. As a combined instruction, that only needs to be done once.
+Other parts of the guard tests also become redundant in the combined
+instruction and can be omitted. For example, the `is_integer/1` type
+test will also succeed if its argument is a **bignum** (an integer
+that does not fit in a machine word). Clearly, a bignum will fall well
+outside the range 0 through 1023, so if the argument is not a small
+integer, the combined guard test will fail immediately.
+
+With those and some other simplifications, we end up with the following
+native instructions:
+
+```nasm
+# is_int_in_range_fScc
+    mov rax, qword ptr [rbx]
+    sub rax, 15
+    test al, 15
+    short jne label_19
+    cmp rax, 16368
+    short ja label_19
+```
+
+The first instruction fetches the value of `{x,0}` to the CPU
+register `rax`:
+
+```nasm
+    mov rax, qword ptr [rbx]
+```
+
+The next instruction subtracts the [tagged value][tags] of the lower
```

Comment:
[tag value][tags]

Reply:
It happens to subtract the tag when the lower bound is 0, but in general it subtracts the tagged value for the lower bound (31 if the lower bound is 1, for example). So I think it should be "tagged value".

Reply:
My bad.  I see it now